File: webkit.c
// 6 october 2014

#include "gtk_unix.h"
#include "_cgo_export.h"

#define MSXHELP "ms-xhelp"

static void webViewLoadMSXHELP(WebKitURISchemeRequest *request, gpointer data)
{
	mainWindowDoFollowLink((void *) data, (char *) webkit_uri_scheme_request_get_uri(request));
}

WebKitWebView *newWebView(void *gomw)
{
	WebKitWebView *wv;

	wv = WEBKIT_WEB_VIEW(webkit_web_view_new());

	// no cache
	webkit_web_context_set_cache_model(webkit_web_view_get_context(wv), WEBKIT_CACHE_MODEL_DOCUMENT_VIEWER);

	// set up MSHI URIs
	webkit_web_context_register_uri_scheme(webkit_web_view_get_context(wv), MSXHELP,
		webViewLoadMSXHELP, (gpointer) gomw, NULL);

	return wv;
}

File: webkit.m
// 6 october 2014
#import <Cocoa/Cocoa.h>
#import <WebKit/WebKit.h>
#import "cocoa_darwin.h"
#import "_cgo_export.h"

#define MSXHELP "ms-xhelp"

/*TODO
static void webViewLoadMSXHELP(WebKitURISchemeRequest *request, gpointer data)
{
	mainWindowDoFollowLink((void *) data, (char *) webkit_uri_scheme_request_get_uri(request));
}
*/

@interface webviewDelegate : NSObject<WebFrameLoadDelegate>
@end

@implementation webviewDelegate

- (void)webView:(WebView *)wv didFailProvisionalLoadWithError:(NSError *)err forFrame:(WebFrame *)frame
{
	NSAlert *alert;

	alert = [NSAlert alertWithError:err];
	[alert runModal];
}

@end

//#define get(x) ((WebView *) [((NSScrollView *) x) documentView])
#define get(x) ((WebView *) (x))

void webviewNavigate(goid webview, char *path)
{
	WebView *wv = get(webview);
	NSString *spath;
	NSURL *url;
	NSURLRequest *req;

	// TODO clear all CSS
	spath = [NSString stringWithUTF8String:path];
	free(path);
	url = [NSURL fileURLWithPath:spath];
	req = [NSURLRequest requestWithURL:url
		cachePolicy:NSURLRequestReloadIgnoringCacheData
		timeoutInterval:INFINITY];			// TODO verify
	[[wv mainFrame] loadRequest:req];
}

goid newWebView(void)
{
	WebView *wv;

	// TODO verify against Interface Builder
	wv = [[WebView alloc] initWithFrame:NSZeroRect frameName:nil groupName:nil];
	[wv setShouldCloseWithWindow:YES];

	// no cache
	// TODO disable the cache or switch to document viewer cache model

	[wv setFrameLoadDelegate:[webviewDelegate new]];

	// set up MSHI URIs
	// TODO set up MSHI URIs

	return wv;
}
